TicketBookingSystemNew.sol

```diff
@@ -0,0 +1,357 @@
+
+contract TicketBookingSystem {
+    address payable public salesManager;
+    address public ticketBuyer;
+    string constant _seatViewLink = "https://seatplan.com/";
+    uint256 numberOfSeatsPerRow = 10;
+    uint256 numberOfRows = 2;
+    uint256 totalNumberOfSeats = numberOfSeatsPerRow * numberOfRows;
+    uint256 availableTickets = totalNumberOfSeats;
+    uint256 showDate;
+    string showTitle;
+    uint256 price;
+    ShowStatus showStatus;
+    Ticket ticketInstance;
+    Poster posterInstance;
+
+    //this mapping stores amount to be refunded against each address if show gets cancelled
+    mapping(address => uint256) amountToBeRefunded;
+
+    //this mapping stores a Seat struct against each seat index (1-60)
+    mapping(uint256 => Seat) public seats;
+
+    //this mapping stores an array of ticketIds bought by each address
+    //each address can purchase multiple tickets
+    mapping(address => uint256[]) ticketsSold;
+
+    enum ShowStatus {
+        Scheduled,
+        Over,
+        Cancelled
+    }
+
+    constructor(
+        string memory _showTitle,
+        uint256 _showDate,
+        uint256 _ticketPrice
+    ) payable {
+        bytes memory tempEmptyStringTest = bytes(_showTitle);
+
+        //check if show title is empty string
+        require(tempEmptyStringTest.length != 0, "Show title can not be empty");
+
+        //check if show date is in past
+        require(
+            _showDate >= block.timestamp,
+            "Show can not be scheduled in past"
+        );
+        //check if ticket price is > 0
+        require(_ticketPrice > 0, "Ticket price must be greater than 0");
+
+        //create new instance of Ticket token
+        ticketInstance = new Ticket();
+        posterInstance = new Poster();
+
+        //set show date, title and price (assume each ticket is same price)
+        showDate = _showDate;
+        showTitle = _showTitle;
+        price = _ticketPrice;
+
+        //set sales manager account address
+        salesManager = payable(msg.sender);
+
+        //for each row and for each seat within row, initialize a seat struct
+        for (uint256 row = 1; row <= numberOfRows; row++) {
+            for (uint256 seat = 1; seat <= numberOfSeatsPerRow; seat++) {
+                seats[seatIndex(row, seat)] = Seat({
+                    isTaken: false,
+                    showTitle: _showTitle,
+                    showDate: _showDate,
+                    ticketPrice: _ticketPrice,
+                    seatRow: row,
+                    seatNumber: seat,
+                    seatViewLink: _seatViewLink
+                });
+            }
+        }
+    }
+
+    /*
+    this function returns the seat index from 1 to rows*seatsPerRow, so in this case
+    it will return an index from 1 to 60. e.g., if row is 2 and seat is 13, index will
+    be 33 
+    */
+    function seatIndex(uint256 row, uint256 seat) private returns (uint256) {
+        return numberOfSeatsPerRow * (row - 1) + seat;
+    }
```

Comment:
Should this also be changed to the correct number for nRows * nSeatsPerRow ?

Reply:
yes